File: server/src/apierror.rs
use actix_web::{HttpResponse, ResponseError};
use reqwest::StatusCode;
use std::fmt::Display;

pub type Result<T> = std::result::Result<T, ApiError>;
pub type Response = Result<HttpResponse>;

#[derive(Debug)]
pub enum ApiError {
    NotFound,
    InternalServerError,
    BadJwt,
    DestInvalid,
    CodeInvalid,
    CodeTaken,
    CodeReserved,
}
impl ApiError {
    pub fn status(&self) -> StatusCode {
        match self {
            ApiError::NotFound => StatusCode::NOT_FOUND,
            ApiError::InternalServerError => StatusCode::INTERNAL_SERVER_ERROR,
            ApiError::BadJwt => StatusCode::UNAUTHORIZED,
            ApiError::CodeTaken => StatusCode::CONFLICT,
            ApiError::CodeReserved => StatusCode::CONFLICT,
            _ => StatusCode::BAD_REQUEST,
        }
    }
    pub fn code(&self) -> u16 {
        match self {
            ApiError::NotFound => 404,
            ApiError::InternalServerError => 500,
            ApiError::BadJwt => 10000,
            ApiError::DestInvalid => 10001,
            ApiError::CodeInvalid => 10002,
            ApiError::CodeTaken => 10003,
            ApiError::CodeReserved => 10004,
        }
    }
}
impl Display for ApiError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{}",
            match self {
                ApiError::NotFound => "not found",
                ApiError::InternalServerError => "internal server error",
                ApiError::BadJwt => "Invalid or missing jwt token",
                ApiError::DestInvalid => "Destination invalid",
                ApiError::CodeInvalid => "Code invalid",
                ApiError::CodeTaken => "Code in use",
                ApiError::CodeReserved => "Code is reserved",
            }
        )
    }
}
impl ResponseError for ApiError {
    fn status_code(&self) -> StatusCode {
        self.status()
    }
    fn error_response(&self) -> HttpResponse {
        HttpResponse::build(self.status()).body(
            String::new()
                + "{\"error\":true,\"message\":\""
                + &self.to_string()
                + "\",\"code\":"
                + &self.code().to_string()
                + "}",
        )
    }
}

File: server/src/lib/ApiError.ts
import express, { Request, Response } from "express";

export class ApiError extends Error {
    private static mkTypes(types: ApiErrorType[]) {
        const map = new Map<number, ApiErrorType>();
        types.forEach(t => map.set(t.id, t));

        return map;
    }
    public static Types = ApiError.mkTypes([
        // basic
        { id: 404, code: 404, name: "not found" },

        // custom
        { id: 10000, name: "Invalid or missing jwt token", code: 403 },
        { id: 10001, name: "Destination invalid" },
        { id: 10002, name: "Code invalid" },
        { id: 10003, name: "Code in use", code: 409 },
        { id: 10004, name: "Code is reserved", code: 409 },
    ]);

    public static Handler(
        err: ApiError | Error,
        req: Request,
        res: Response,
        next: express.NextFunction
    ) {
        if (res.headersSent) return next(err);

        // console.error(err);

        const d: any = { error: true, message: err.message };

        if (err instanceof ApiError) {
            res.status(err.code);
            d.code = err.id;
        }

        res.json(d);
    }

    message: string;
    id: number;
    code: number;

    constructor(id: number) {
        super();

        const type = ApiError.Types.get(id);
        if (!type) throw new Error(`${id} is not a valid API error code`);

        this.message = type.name;
        this.id = id;
        this.code = type.code || 400;
    }
}
export default ApiError;

export interface ApiErrorType {
    name: string;
    id: number;
    code?: number;
}
